micrometer-metrics/micrometer
# Prometheus histogram exemplars

fixes #2812
closes #2672

# How to test/use this

There are various ways to try this, here's what you need to do and should see:
- You need to pass a `SpanContextSupplier` implementation to the `PrometheusMeterRegistry`
- You need to use a `Counter` or a `Timer` or a `DistributionSummary` (`publishPercentileHistogram` or `serviceLevelObjectives` must be enabled)
- You need to call `scrape` using the `openmetrics-text` format
- You should see something like this: `test_counter_total 1.0 # {span_id="1",trace_id="2"} 1.0 1647993571.375`  
The part after the `#` belongs to the Exemplar: spanId, traceID, captured value (incrementing the counter by 1), timestamp.

## Run `PrometheusExemplarsSample`

This sample does everything for you (using a "fake" `SpanContextSupplier` implementation), check the output and look for `trace_id` and `span_id`, e.g.:
```
test_counter_total 1.0 # {span_id="1",trace_id="2"} 1.0 1647993571.375
```
or
```
test_timer_seconds_bucket{le="0.001"} 1.0 # {span_id="3",trace_id="4"} 0.001 1647993571.398
```

## Check the added verifications to `PrometheusMeterRegistryTest`

The tests added in this PR are pretty similar to `PrometheusExemplarsSample` but you can also see what to expect in the verification steps. 

## Run `PrometheusSample`

This is a full-fledged example with a real Spring-Boot application using a real Distributed Tracing solution (Spring Cloud Sleuth).

After you start the app, you need to generate some traffic. If you record something using a `Counter` or a `Timer` or a `DistributionSummary` (`publishPercentileHistogram` or `serviceLevelObjectives` must be enabled), and that recording is in the scope of Sleuth, you should also see an Exemplar created to the involved time series.  
Since Spring Boot does this for you out of the box, you should see the exemplars if you hit any method of the controller in the sample, e.g.:
```
http :8080/api/people
```

Doing this should already record the Exemplar for you, you should see it if you scrape the Prometheus endpoint using the `openmetrics-text` format:
```
http :8080/actuator/prometheus 'Accept: application/openmetrics-text; version=1.0.0' | grep trace_id
```

## `PrometheusSample` with Prometheus, Grafana and Zipkin

You can see this in action using Prometheus, Grafana and Zipkin, here's how to do it:
1. Run `PrometheusSample`
2. Go to https://github.com/jonatan-ivanov/local-services and start Prometheus and Zipkin using its docker-compose files
(clone, cd into the prometheus/zipkin dirs, and `docker compose up`)
3. Hit the "person" endpoint (I prepared a `Counter` for you there): `http :8080/api/person/1`
4. Go to your local Prometheus instance ([link](http://localhost:9090/graph?g0.expr=person_requests_total&g0.tab=0&g0.stacked=0&g0.show_exemplars=1&g0.range_input=1h)), query `person_requests_total` and click "Show Exemplars":  
The Exemplar is a little diamond on your graph that shows you the traceId and the spanId if you hover over it
6. Go to your local Grafana instance and check the "Exemplars" dashboard ([link](http://localhost:3000/d/fTL0SuI7k/exemplars?orgId=1&refresh=5s)):  
The Exemplar is a little diamond on your graph that shows you the traceId and the spanId if you hover over it, also if you press the "Go to" button next to the traceId, it will open that particular Trace in Zipkin
